File: python/src/labstacknetworks/resources/provider.py
from __future__ import annotations

from typing import Any, Mapping, Optional, Sequence


class ProviderResource:
    """Provider API — https://integration.labstack.in/api-docs/provider

    Speciality discovery, provider search, appointment booking, document
    management and video consultations.
    """

    def __init__(self, client):
        self._client = client

    # ---- admin ----------------------------------------------------------

    def get_status(self) -> dict:
        return self._client.request("GET", "/api/v1.0/admin/status")

    def get_balance(self) -> dict:
        return self._client.request("GET", "/api/v1.0/admin/getBalance")

    # ---- network / availability -------------------------------------------

    def get_provider_details(self, ls_provider_id: str, procedure_id: Optional[str] = None) -> dict:
        return self._client.request(
            "GET", "/api/v1.0/network/getProviderDetails", query={"lsProviderId": ls_provider_id, "procedureId": procedure_id}
        )

    def get_specialities(self, provider_type: str, tags: Optional[str] = None) -> dict:
        return self._client.request(
            "GET", "/api/v1.0/availability/getSpecialities", query={"providerType": provider_type, "tags": tags}
        )

    def get_providers(self, provider_type: str, appointment_type: str, **params: Any) -> dict:
        query = {"providerType": provider_type, "appointmentType": appointment_type, **params}
        return self._client.request("GET", "/api/v1.0/availability/getProviders", query=query)

    def get_provider_tags(self, **params: Any) -> dict:
        return self._client.request("GET", "/api/v1.0/availability/getProviderTags", query=params)

    def get_provider_tag_categories(self, **params: Any) -> dict:
        return self._client.request("GET", "/api/v1.0/availability/getProviderTagCategories", query=params)

    def get_procedures(self, **params: Any) -> dict:
        return self._client.request("GET", "/api/v1.0/availability/getProcedures", query=params)

    def get_procedure(self, procedure_id: str, **params: Any) -> dict:
        query = {"procedureId": procedure_id, **params}
        return self._client.request("GET", "/api/v1.0/availability/getProcedure", query=query)

    def get_appointment_slots(self, provider_type: str, appointment_type: str, date: str, **params: Any) -> dict:
        query = {"providerType": provider_type, "appointmentType": appointment_type, "date": date, **params}
        return self._client.request("GET", "/api/v1.0/availability/getAppointmentSlots", query=query)

    # ---- appointments -------------------------------------------------------

    def book_appointment(self, **body: Any) -> dict:
        """Required: appointment_type, appointment, name, email, mobile, gender
        (pass as appointmentType, appointment, name, email, mobile, gender)."""
        return self._client.request("POST", "/api/v1.0/appointment/bookAppointment", body=body)

    def get_appointments(self, **params: Any) -> dict:
        return self._client.request("GET", "/api/v1.0/appointment/getAppointments", query=params)

    def get_appointment_details(self, appointment_id: int) -> dict:
        return self._client.request("GET", "/api/v1.0/appointment/getAppointmentDetails", query={"appointmentId": appointment_id})

    def cancel_appointment(self, appointment_id: int, reason: Optional[str] = None) -> dict:
        return self._client.request(
            "GET", "/api/v1.0/appointment/cancelAppointment", query={"appointmentId": appointment_id, "reason": reason}
        )

    def reschedule_appointment(self, appointment_id: int, appointment_time: str) -> dict:
        return self._client.request(
            "GET",
            "/api/v1.0/appointment/rescheduleAppointment",
            query={"appointmentId": appointment_id, "appointmentTime": appointment_time},
        )

    def raise_escalation(
        self,
        appointment_id: int,
        complaint: str,
        severity: str,
        description: Optional[str] = None,
        notes: Optional[str] = None,
    ) -> dict:
        return self._client.request(
            "POST",
            "/api/v1.0/appointment/escalation",
            body={
                "appointmentId": appointment_id,
                "complaint": complaint,
                "severity": severity,
                "description": description,
                "notes": notes,
            },
        )

    def test_webhook(self, appointment_id: int, appointment_status: Optional[str] = None) -> dict:
        """Trigger a test appointment webhook (sandbox/testing use)."""
        return self._client.request(
            "GET",
            "/api/v1.0/appointment/testWebhook",
            query={"appointmentId": appointment_id, "appointmentStatus": appointment_status},
        )

    # ---- documents & meeting -------------------------------------------------

    def upload_document(self, appointment_id: int, documents: Sequence[Mapping[str, str]]) -> dict:
        return self._client.request(
            "POST",
            "/api/v1.0/appointment/uploadDocument",
            body={"appointmentId": appointment_id, "documents": list(documents)},
        )

    def get_document(self, document_id: int) -> dict:
        return self._client.request("GET", "/api/v1.0/appointment/getDocument", query={"documentId": document_id})

    def delete_document(self, **params: Any) -> dict:
        """Deprecated: currently disabled server-side, always responds 501 Not Implemented."""
        return self._client.request("DELETE", "/api/v1.0/appointment/deleteDocument", query=params)

    def delete_all_documents(self, **params: Any) -> dict:
        """Deprecated: currently disabled server-side, always responds 501 Not Implemented."""
        return self._client.request("DELETE", "/api/v1.0/appointment/deleteAllDocuments", query=params)

    def get_prescription_pdf(self, appointment_id: int) -> dict:
        return self._client.request("GET", "/api/v1.0/appointment/getPrescriptionPdf", query={"appointmentId": appointment_id})

    def get_authenticated_meeting_link(self, appointment_id: int, participant_type: Optional[str] = None) -> dict:
        return self._client.request(
            "GET",
            "/api/v1.0/appointment/getAuthenticatedMeetingLink",
            query={"appointmentId": appointment_id, "participantType": participant_type},
        )

File: go/provider.go
package labstacknetworks

// ProviderResource covers the Provider API —
// https://integration.labstack.in/api-docs/provider
// Speciality discovery, provider search, appointment booking, document
// management and video consultations.
type ProviderResource struct {
	client *Client
}

// ---- admin ----------------------------------------------------------

func (p *ProviderResource) GetStatus() (map[string]any, error) {
	return p.client.request("GET", "/api/v1.0/admin/status", nil, nil)
}

func (p *ProviderResource) GetBalance() (map[string]any, error) {
	return p.client.request("GET", "/api/v1.0/admin/getBalance", nil, nil)
}

// ---- network / availability -------------------------------------------

func (p *ProviderResource) GetProviderDetails(lsProviderID, procedureID string) (map[string]any, error) {
	q := newQuery().set("lsProviderId", lsProviderID).set("procedureId", procedureID).build()
	return p.client.request("GET", "/api/v1.0/network/getProviderDetails", q, nil)
}

func (p *ProviderResource) GetSpecialities(providerType, tags string) (map[string]any, error) {
	q := newQuery().set("providerType", providerType).set("tags", tags).build()
	return p.client.request("GET", "/api/v1.0/availability/getSpecialities", q, nil)
}

func (p *ProviderResource) GetProviders(providerType, appointmentType string, params map[string]any) (map[string]any, error) {
	q := toQuery(params)
	q.Set("providerType", providerType)
	q.Set("appointmentType", appointmentType)
	return p.client.request("GET", "/api/v1.0/availability/getProviders", q, nil)
}

func (p *ProviderResource) GetProviderTags(params map[string]any) (map[string]any, error) {
	return p.client.request("GET", "/api/v1.0/availability/getProviderTags", toQuery(params), nil)
}

func (p *ProviderResource) GetProviderTagCategories(params map[string]any) (map[string]any, error) {
	return p.client.request("GET", "/api/v1.0/availability/getProviderTagCategories", toQuery(params), nil)
}

func (p *ProviderResource) GetProcedures(params map[string]any) (map[string]any, error) {
	return p.client.request("GET", "/api/v1.0/availability/getProcedures", toQuery(params), nil)
}

func (p *ProviderResource) GetProcedure(procedureID string, params map[string]any) (map[string]any, error) {
	q := toQuery(params)
	q.Set("procedureId", procedureID)
	return p.client.request("GET", "/api/v1.0/availability/getProcedure", q, nil)
}

func (p *ProviderResource) GetAppointmentSlots(providerType, appointmentType, date string, params map[string]any) (map[string]any, error) {
	q := toQuery(params)
	q.Set("providerType", providerType)
	q.Set("appointmentType", appointmentType)
	q.Set("date", date)
	return p.client.request("GET", "/api/v1.0/availability/getAppointmentSlots", q, nil)
}

// ---- appointments -------------------------------------------------------

// BookAppointment requires: appointmentType, appointment, name, email, mobile, gender.
// See spec/provider.openapi.json at the repo root for the full field list.
func (p *ProviderResource) BookAppointment(body map[string]any) (map[string]any, error) {
	return p.client.request("POST", "/api/v1.0/appointment/bookAppointment", nil, body)
}

func (p *ProviderResource) GetAppointments(params map[string]any) (map[string]any, error) {
	return p.client.request("GET", "/api/v1.0/appointment/getAppointments", toQuery(params), nil)
}

func (p *ProviderResource) GetAppointmentDetails(appointmentID int64) (map[string]any, error) {
	q := newQuery().set("appointmentId", appointmentID).build()
	return p.client.request("GET", "/api/v1.0/appointment/getAppointmentDetails", q, nil)
}

func (p *ProviderResource) CancelAppointment(appointmentID int64, reason string) (map[string]any, error) {
	q := newQuery().set("appointmentId", appointmentID).set("reason", reason).build()
	return p.client.request("GET", "/api/v1.0/appointment/cancelAppointment", q, nil)
}

func (p *ProviderResource) RescheduleAppointment(appointmentID int64, appointmentTime string) (map[string]any, error) {
	q := newQuery().set("appointmentId", appointmentID).set("appointmentTime", appointmentTime).build()
	return p.client.request("GET", "/api/v1.0/appointment/rescheduleAppointment", q, nil)
}

func (p *ProviderResource) RaiseEscalation(appointmentID int64, complaint, severity, description, notes string) (map[string]any, error) {
	body := map[string]any{
		"appointmentId": appointmentID,
		"complaint":     complaint,
		"severity":      severity,
		"description":   description,
		"notes":         notes,
	}
	return p.client.request("POST", "/api/v1.0/appointment/escalation", nil, body)
}

// TestWebhook triggers a test appointment webhook (sandbox/testing use).
func (p *ProviderResource) TestWebhook(appointmentID int64, appointmentStatus string) (map[string]any, error) {
	q := newQuery().set("appointmentId", appointmentID).set("appointmentStatus", appointmentStatus).build()
	return p.client.request("GET", "/api/v1.0/appointment/testWebhook", q, nil)
}

// ---- documents & meeting -------------------------------------------------

func (p *ProviderResource) UploadDocument(appointmentID int64, documents []map[string]string) (map[string]any, error) {
	body := map[string]any{"appointmentId": appointmentID, "documents": documents}
	return p.client.request("POST", "/api/v1.0/appointment/uploadDocument", nil, body)
}

func (p *ProviderResource) GetDocument(documentID int64) (map[string]any, error) {
	q := newQuery().set("documentId", documentID).build()
	return p.client.request("GET", "/api/v1.0/appointment/getDocument", q, nil)
}

// DeleteDocument is deprecated: currently disabled server-side, always
// responds 501 Not Implemented.
func (p *ProviderResource) DeleteDocument(params map[string]any) (map[string]any, error) {
	return p.client.request("DELETE", "/api/v1.0/appointment/deleteDocument", toQuery(params), nil)
}

// DeleteAllDocuments is deprecated: currently disabled server-side, always
// responds 501 Not Implemented.
func (p *ProviderResource) DeleteAllDocuments(params map[string]any) (map[string]any, error) {
	return p.client.request("DELETE", "/api/v1.0/appointment/deleteAllDocuments", toQuery(params), nil)
}

func (p *ProviderResource) GetPrescriptionPdf(appointmentID int64) (map[string]any, error) {
	q := newQuery().set("appointmentId", appointmentID).build()
	return p.client.request("GET", "/api/v1.0/appointment/getPrescriptionPdf", q, nil)
}

func (p *ProviderResource) GetAuthenticatedMeetingLink(appointmentID int64, participantType string) (map[string]any, error) {
	q := newQuery().set("appointmentId", appointmentID).set("participantType", participantType).build()
	return p.client.request("GET", "/api/v1.0/appointment/getAuthenticatedMeetingLink", q, nil)
}
